go-metadata/istio.go

```diff
@@ -0,0 +1,132 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package istiometadata
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net"
+	"net/http"
+	"net/url"
+	"time"
+)
+
+// ConnectionMetadata returns the metadata about a connection
+type ConnectionMetadata struct {
+	Identity string `json:"identity"`
+}
+
+// metadataClient provides a default client for all metadata requests
+var metadataClient = http.Client{
+	Timeout: time.Second,
+}
+
+// metadataContextKey is the key the metadata handler will store connection metadata in
+var metadataContextKey = &contextKey{}
+
+type contextKey struct {
+	name string
+}
+
+func (k *contextKey) String() string { return "istio metadata context value" }
+
+func metadataServerURL() *url.URL {
+	u, err := url.Parse("http://169.254.169.111")
```

Comment:
For google there is a well-known env variable that may contain the MDS address.

I would use something like AMBIENT_MDS - the user may have multiple MDS servers ( platform, etc).
We should also consider using a Service (mds.istio-system.svc or ambientmds.istio-system.svc), with node affinity - it may be possible to run ztunnel without CNI ( using SOCKS for egress ), this would allow accessing ztunnel mds without magic.